[nmcli] Fix issue with argument conversion for run_command

marcin-sucharski opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Changes in this PR are fixing the issue where nmcli module would throw an error when executing `self.module.run_command`. Arguments passed to `run_command` are expected to be a list of strings (or single string). In the case of `nmcli` module, sometimes integer value was present in the list which caused `run_command` to fail.

I've also noticed that there are references to undefined variable `self.vxlanid`. I've replaced them with `self.vxlan_id`.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
nmcli

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->

I've noticed this bug when running following task:
```
- name: Add bridge connection for virtual machines.
  nmcli:
    type: bridge
    autoconnect: no
    conn_name: '{{ qemu_network_manager_bridge_name }}'
    ifname: '{{ qemu_network_manager_bridge_name }}'
    ip4: '{{ qemu_network_manager_bridge_ip4 }}'
    stp: no
    state: present
```
And it gave me following error:
```
TASK [qemu : Add bridge connection for virtual machines.] *********************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: expected str, bytes or os.PathLike object, not int
fatal: [host]: FAILED! => changed=false 
  module_stderr: |-
    Traceback (most recent call last):
      File "<stdin>", line 114, in <module>
      File "<stdin>", line 106, in _ansiballz_main
      File "<stdin>", line 49, in invoke_module
      File "/usr/lib64/python3.6/imp.py", line 235, in load_module
        return load_source(name, filename, file)
      File "/usr/lib64/python3.6/imp.py", line 170, in load_source
        module = _exec(spec, sys.modules[name])
      File "<frozen importlib._bootstrap>", line 618, in _exec
      File "<frozen importlib._bootstrap_external>", line 678, in exec_module
      File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
      File "/tmp/ansible_nmcli_payload_oi1b0db1/__main__.py", line 1549, in <module>
      File "/tmp/ansible_nmcli_payload_oi1b0db1/__main__.py", line 1527, in main
      File "/tmp/ansible_nmcli_payload_oi1b0db1/__main__.py", line 1425, in modify_connection
      File "/tmp/ansible_nmcli_payload_oi1b0db1/__main__.py", line 682, in execute_command
      File "/tmp/ansible_nmcli_payload_oi1b0db1/ansible_nmcli_payload.zip/ansible/module_utils/basic.py", line 2477, in run_command
      File "/tmp/ansible_nmcli_payload_oi1b0db1/ansible_nmcli_payload.zip/ansible/module_utils/basic.py", line 2477, in <listcomp>
      File "/usr/lib/python-exec/python3.6/../../../lib64/python3.6/posixpath.py", line 281, in expandvars
        path = os.fspath(path)
    TypeError: expected str, bytes or os.PathLike object, not int
  module_stdout: ''
  msg: |-
    MODULE FAILURE
    See stdout/stderr for the exact error
  rc: 1
```

I'm using Ansible 2.8.0.

bcoca reviewed (changes requested):
first, never use `str` use `to_` functions we provide (in this case `to_text` is the appropriate one), 2nd i would argue that you should not do this on each value, but ensure that all elements in the list in run_command are strings.

marcin-sucharski commented:
@bcoca Thanks for your feedback!

I've replaced `str` with `to_text`. I've decided to do this only on the values which are expected to be non-string and will correctly convert using `to_text` (boolean, for example, won't convert correctly).

For this reason (correctness) I think this should be done for each value, not for all of them. However, if it is preffered, I will change the implementation to one that will do this for all arguments.

Also I've added `to_text` with `nonstring='strict'` to `execute_command` in nmcli module to make sure that all arguments passed to `self.module.run_command` are strings (`nonstring='strict'` will protect from unexpected conversions). I tried to add this conversion with strict to `module.run_command`, but some tests started to fail and I do not know Ansible codebase good enough to fix this in reasonable time.

Akasurde commented on lib/ansible/modules/net_tools/nmcli.py:
```diff
@@ -1002,7 +1002,7 @@ def modify_connection_ethernet(self, conn_type='ethernet'):
             'ipv4.dns-search': self.dns4_search,
             'ipv6.dns-search': self.dns6_search,
-            '802-3-ethernet.mtu': self.mtu,
+            '802-3-ethernet.mtu': str(self.mtu),
```
How about adding `to_text(value)` in next loop where we are iterating on options ? 

```python
        for key, value in options.items():
            if value is not None:
                if key == '802-3-ethernet.mtu' and conn_type != 'ethernet':
                    continue
                cmd.extend([key, to_text(value)])
```

cinpol replied:
Importing the `to_text` function
```python
from ansible.module_utils._text import to_text
```
and replacing
```python
        for key, value in options.items():
            if value is not None:
                cmd.extend([key, value])
```
with
```python
        for key, value in options.items():
            if value is not None:
                cmd.extend([key, to_text(value)])
```
did solve issue https://github.com/ansible/ansible/issues/59095 for me. I however don't know what the consequences are.

marcin-sucharski replied:
@Akasurde I've made a mistake during previous rebase and didn't notice that change of `str` to `to_text` was lost. I've fixed that right now.

As mentioned in previous comment - the converion is done for every parameter explicitly on purpose.
I've decided to do this only on the values which are expected to be non-string and will correctly convert using to_text (boolean, for example, won't convert correctly).
For this reason (correctness) I think this should be done for each value, not for all of them.

However, if it is preffered, I will change the implementation to one that will do this for all arguments (in the loop). Just let me know.

@polatsinan Does the issue persist when applying current changes in the PR?

bcoca replied:
@marcin-sucharski check against current development, I've merged stringification at run_command so you might not need to do this anymore.

cinpol replied:
@bcoca I have cloned the devel branche and the issue still exists:
```
ansible 2.9.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/spolat/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/spolat/python/ansible/local/lib/python2.7/site-packages/ansible
  executable location = /home/spolat/python/ansible/bin/ansible
  python version = 2.7.15+ (default, Nov 27 2018, 23:36:35) [GCC 7.3.0]
```

```
TypeError: argument of type 'int' is not iterable
```

@marcin-sucharski: Applying current changes in the PR fixes the issue.

bcoca replied:
that is not same issue as original above, can you provide -vvv output?

Commit:
Fix issue with argument conversion for run_command

In the nmcli module arguments in list passed to run_command sometimes
were of type int. This caused issue in the run_command which resulted
in exception being thrown.

marcin-sucharski commented:
@bcoca The issue is still present in the release v2.8.4. Is there any way I can help to move this forward?

numas commented:
Got hit by this issue once again, creating a bridge. Fixed it manually by editing bridge functions in nmcli.py using @Akasurde suggestion:
```python
        for key, value in options.items():
            if value is not None:
                cmd.extend([key, to_text(value)])
```
Not to sound rude, but this should really move forward. I understand the need for "correctness" in code, but the decision of converting all values to string or only specific values, should not be more important than fixing a currently broken module.

kbrowder commented:
Seems to still be broken, I think nmcli doesn't work for anyone right?  Or is it just python3 people?

martingolas commented:
Yep, still broken on CentOS 7.7.1908 and base Python 2.7.5.

cppmonkey commented on lib/ansible/modules/net_tools/nmcli.py:
```diff
@@ -1221,9 +1225,9 @@ def modify_connection_vxlan(self):
             cmd.append(self.ifname)
         else:
-            cmd.append('vxlan%s' % self.vxlanid)
+            cmd.append('vxlan%s' % to_text(self.vxlan_id))
```
Thanks this PR really helped me out!

These format statements don't need the to_text, as it will convert the variables to a string
https://docs.python.org/2/library/stdtypes.html#string-formatting

```
$ ipython
Python 3.8.2 (default, Feb 28 2020, 00:00:00) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: var = 10                                                                                                                                                                                        

In [2]: var                                                                                                                                                                                             
Out[2]: 10

In [3]: type(var)                                                                                                                                                                                       
Out[3]: int

In [4]: formatted = 'vlan%s' % var                                                                                                                                                                      

In [5]: formatted                                                                                                                                                                                       
Out[5]: 'vlan10'

In [6]: type(formatted)                                                                                                                                                                                 
Out[6]: str
```

Seems a shame to change everything to a String.
Seen a number of nice PRs for nmcli which haven't been merged and one impacts Fedora 31/32, RHEL8 and CentOS8.

Best of luck

Akasurde commented:
Migrated to https://github.com/ansible-collections/community.general/pull/477
